el-patch-template.el

```diff
@@ -0,0 +1,433 @@
+                                          (list new-match))
+                                  next-step-fn
+                                  table literal in-string))))
+          (el-patch-template--process (car form) (car template)
+                                      nil consp-next-step table
+                                      literal in-string))))
+     ((null template) ;; nothing else to match
+      (funcall next-step-fn match form))
+     ((and in-string (stringp form)
+           (stringp template)
+           (string-prefix-p template form))
+      ;; Allow matching a substring with IN-STRING
+      (funcall next-step-fn (append match template)
+               (when (> (length form) (length template))
+                 (substring form (length template)))))
+     ((or (member template '(...)) (equal template form))
+      ;; A Complete match.
+      (funcall next-step-fn (append match form) nil))
+     (t
+      (or (when-let ((symbol (gethash template table)))
+            (el-patch-template--process form
+                                        (or
+                                         ;; The previous resolution
+                                         (cdr symbol)
+                                         ;; The template-value
+                                         (car symbol))
+                                        nil
+                                        (lambda (new-match
+                                                 remainder-form)
+                                          ;; First put the match in
+                                          ;; the table
+                                          (puthash template
+                                                   (cons
+                                                    (car symbol)
+                                                    new-match)
+                                                   table)
+                                          ;; Then process the next
+                                          ;; step, adding the template
+                                          ;; to the match
+                                          (funcall next-step-fn
+                                                   (if match
+                                                       (cons match
+                                                             template)
+                                                     template)
+                                                   remainder-form))
+                                        table literal in-string))
+          (throw 'no-match nil))))))
+
+(defun el-patch-template--match-p (form template)
+  "Matches TEMPLATE to FORM. TEMPLATE may contain `...' which
+greedily match any number of form. Match is succesful if a
+partial list of FORM, starting from the beginning matches
+TEMPLATE. The return value is the number of forms in FORM which
+match template or nil if a match is not possible."
```

Comment:
This is a fast version of `el-patch-template--process` that does not expect `el-patch-*` forms but does handle `...`. It is used as a fast matching function, so it must be called with a resolved template. When a match is successful, the function `el-patch-template--process` is then called with the original template (which includes `el-patch-*`).